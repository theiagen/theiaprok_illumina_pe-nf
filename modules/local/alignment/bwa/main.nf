process BWA_MEM {
    tag "$meta.id"
    label "process_low"
    
    container "us-docker.pkg.dev/general-theiagen/staphb/ivar:1.3.1-titan"

    input:
    tuple val(meta), path(reads), path(reference)

    output:
    tuple val(meta), path('*.sorted.bam')    , emit: bam
    tuple val(meta), path('*.sorted.bam.bai'), emit: bai
    tuple val(meta), path('*_R1.fastq.gz')  , emit: fastq_1
    tuple val(meta), path('*_R2.fastq.gz')  , emit: fastq_2, optional: true
    tuple val(meta), path('*_unaligned_R1.fastq.gz'), emit: unaligned_1
    tuple val(meta), path('*_unaligned_R2.fastq.gz'), emit: unaligned_2, optional: true
    path "versions.yml"                      , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def read_group = meta.read_group ? "-R ${meta.read_group}" : "-R '@RG\\tID:${prefix}\\tSM:${prefix}'"
    def reads_command = meta.single_end ? "${reads}" : "${reads[0]} ${reads[1]}"
    
    """
    # Index reference
    bwa index ${reference}

    # Align reads
    bwa mem \\
        -t ${task.cpus} \\
        ${read_group} \\
        ${reference} \\
        ${reads_command} \\
        ${args} | \\
    samtools sort \\
        -@ ${task.cpus} \\
        -o ${prefix}.sorted.bam

    # Create separate BAM for aligned reads only (no secondary/supplementary)
    samtools view \\
        -@ ${task.cpus} \\
        -F 0x904 \\
        -b \\
        -o ${prefix}.sorted.aligned-only.bam \\
        ${prefix}.sorted.bam

    # Create BAM for unaligned reads
    samtools view \\
        -@ ${task.cpus} \\
        -f 4 \\
        -b \\
        -o ${prefix}.sorted.unaligned-reads.bam \\
        ${prefix}.sorted.bam

    # Extract FASTQ files
    if [[ "${meta.single_end}" == "true" ]]; then
        # Single-end reads
        samtools fastq \\
            -@ ${task.cpus} \\
            -F 4 \\
            -0 ${prefix}_R1.fastq.gz \\
            ${prefix}.sorted.aligned-only.bam

        samtools fastq \\
            -@ ${task.cpus} \\
            -f 4 \\
            -0 ${prefix}_unaligned_R1.fastq.gz \\
            ${prefix}.sorted.unaligned-reads.bam
    else
        # Paired-end reads
        samtools fastq \\
            -@ ${task.cpus} \\
            -F 4 \\
            -1 ${prefix}_R1.fastq.gz \\
            -2 ${prefix}_R2.fastq.gz \\
            ${prefix}.sorted.aligned-only.bam

        samtools fastq \\
            -@ ${task.cpus} \\
            -f 4 \\
            -1 ${prefix}_unaligned_R1.fastq.gz \\
            -2 ${prefix}_unaligned_R2.fastq.gz \\
            ${prefix}.sorted.unaligned-reads.bam
    fi

    # Index BAM files
    samtools index ${prefix}.sorted.bam
    samtools index ${prefix}.sorted.unaligned-reads.bam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bwa: \$(echo \$(bwa 2>&1) | sed 's/^.*Version: //; s/Contact:.*\$//')
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.sorted.bam
    touch ${prefix}.sorted.bam.bai
    touch ${prefix}_R1.fastq
    touch ${prefix}_unaligned_R1.fastq
    if [[ "${meta.single_end}" != "true" ]]; then
        touch ${prefix}_R2.fastq
        touch ${prefix}_unaligned_R2.fastq

    fi

    gzip *.fastq

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        bwa: \$(echo \$(bwa 2>&1) | sed 's/^.*Version: //; s/Contact:.*\$//')
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """
}