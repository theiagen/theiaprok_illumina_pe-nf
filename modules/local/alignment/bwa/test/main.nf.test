nextflow_process {
    name "Test Process BWA_MEM"
    script "../main.nf"
    process "BWA_MEM"
    
    test("bwa_mem - test") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz", checkIfExists: true)
                    ]
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {  assert snapshot(
                    process.out.bam,
                    process.out.bai,
                    process.out.fastq_1,
                    process.out.fastq_2,
                    process.out.unaligned_1,
                    process.out.unaligned_2,
                    process.out.versions,
                    ).match() }
            )
        }
    }
    
    test("bwa_mem - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                    ],
                    [
                        file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fasta/test1.contigs.fa.gz", checkIfExists: true)
                    ]
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}