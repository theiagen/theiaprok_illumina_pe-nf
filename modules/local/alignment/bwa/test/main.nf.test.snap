{
    "bwa_mem - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.sorted.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.sorted.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        [
                            "test_sample_R1.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_sample_unaligned_R1.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        [
                            "test_sample_R2.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_sample_unaligned_R2.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_unaligned_R1.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_unaligned_R2.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    "versions.yml:md5,0ce18ad1a5a5d823919cde11e3086dac"
                ],
                "bai": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.sorted.bam.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "bam": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.sorted.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "fastq_1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        [
                            "test_sample_R1.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_sample_unaligned_R1.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "fastq_2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        [
                            "test_sample_R2.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_sample_unaligned_R2.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "unaligned_1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_unaligned_R1.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "unaligned_2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_unaligned_R2.fastq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0ce18ad1a5a5d823919cde11e3086dac"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-20T20:37:05.126000377"
    },
    "bwa_mem - test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.sorted.bam:md5,1b8cc78682f762db5afa4710712160d0"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.sorted.bam.bai:md5,65e4fd3706a8465fcbc2b985e29885fe"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    [
                        "test_sample_R1.fastq.gz:md5,35eb5ebcb6e5ae41d27be2f811d1afbf",
                        "test_sample_unaligned_R1.fastq.gz:md5,5b91c6081dbc33214068798a3df58ca9"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    [
                        "test_sample_R2.fastq.gz:md5,fc2698648a2c795deec1f6d8473599cc",
                        "test_sample_unaligned_R2.fastq.gz:md5,78ab2dd0301ca6cd77d210178f140ab7"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_unaligned_R1.fastq.gz:md5,5b91c6081dbc33214068798a3df58ca9"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_unaligned_R2.fastq.gz:md5,78ab2dd0301ca6cd77d210178f140ab7"
                ]
            ],
            [
                "versions.yml:md5,0ce18ad1a5a5d823919cde11e3086dac"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-20T20:36:52.13622899"
    }
}